Use SDK ILLink analyzer for OOBs for downlevel TFMs

sbomer opened the pull request:
Don't use live ILLink analyzer for OOBs when targeting downlevel TFMs. Should help catch issues like https://github.com/dotnet/runtime/issues/114307.

Note that some OOBs don't target the current TFM - these will not be analyzed using the live analyzer at all. This should be mitigated by our OOB trim step which runs all of the OOB assemblies through the live linker.

sbomer committed:
Use SDK ILLink analyzer for OOBs for downlevel TFMs

ViktorHofer commented on eng/liveILLink.targets:
```diff
@@ -2,7 +2,7 @@
   <!-- NOTE: these targets should be a no-op, unless _RequiresLiveILLink, computed below, is true. -->
 
-  <PropertyGroup>
+  <PropertyGroup Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'">
```
This is incorrect. We have several projects that target i.e. `$(NetCoreAppCurrent)-windows` or similar. You want to use the `IsTargetFrameworkCompatible` function here.

ViktorHofer commented on src/libraries/Directory.Build.targets:
```diff
@@ -135,7 +135,7 @@
   <Import Project="$(RepositoryEngineeringDir)testing\coverage.targets" Condition="'$(EnableRunSettingsSupport)' == 'true' or '$(EnableCoverageSupport)' == 'true'" />
 
-  <Import Project="$(RepositoryEngineeringDir)illink.targets" Condition="'$(IsSourceProject)' == 'true' or '$(ExplicitlyImportCustomILLinkTargets)' == 'true'" />
+  <Import Project="$(RepositoryEngineeringDir)illink.targets" Condition="('$(IsSourceProject)' == 'true' and '$(TargetFramework)' == '$(NetCoreAppCurrent)') or '$(ExplicitlyImportCustomILLinkTargets)' == 'true'" />
```
Same here

ViktorHofer replied:
Looking at all these conditions it might make sense to introduce a deferred property like `ImportCustomILLInkTargets` and condition on that instead. That would allow replacing `ExplicitlyImportCustomILLinkTargets` with it and make this more readable and more extensible.

ViktorHofer commented on src/libraries/Directory.Build.props:
```diff
@@ -49,7 +49,7 @@
     <Nullable Condition="'$(Nullable)' == '' and '$(IsTestProject)' == 'true'">annotations</Nullable>
     <!-- EnableAOTAnalyzer is enabled by default for src projects. -->
-    <EnableAOTAnalyzer Condition="'$(EnableAOTAnalyzer)' == '' and '$(IsSourceProject)' == 'true'">true</EnableAOTAnalyzer>
+    <EnableAOTAnalyzer Condition="'$(EnableAOTAnalyzer)' == '' and '$(IsSourceProject)' == 'true' and $([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0'))">true</EnableAOTAnalyzer>
```
A general principle is to not condition on `TargetFramework` in props files as taht only works in multi-targeting projects for which the inner project build has the TargetFramework available as a global property. For single non-multi targeting builds that's not true and TargetFramework isn't available in props.

Consider moving that to the D.B.targets file.

sbomer committed:
PR feedback

- Use IsTargetFrameworkCompatible
- Move EnableAOTAnalyzer to targets
- Rename ExplicitlyImportCustomILLinkTargets to
  ImportCustomILLinkTargets and infer value

sbomer committed:
Build logic fixes

- Fix typo
- Use IsTargetFrameworkCompatible
- Import illink.targets only for netcoreapp
  To prevent netfx TFMs from attempting to use
  ILLink tasks that aren't available when targeting
  netfx.
- Only use live ILLIink for current TFM

sbomer committed:
Update suppressions in STJ

sbomer committed:
Merge branch 'main' into illinkAnalyzerOob

sbomer committed:
Merge remote-tracking branch 'origin/main' into illinkAnalyzerOob

sbomer committed:
Clean up conditions, ensure projects target AOT supported TFMs

sbomer committed:
Silence AOT compatibility warning for source-build

sbomer committed:
Multitarget Microsoft.XmlSerializer.Generator

ViktorHofer commented on src/libraries/Microsoft.Bcl.AsyncInterfaces/src/Microsoft.Bcl.AsyncInterfaces.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;$(NetFrameworkMinimum);netstandard2.1</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum);netstandard2.1</TargetFrameworks>
```
This looks wrong. A package like Bcl.AsyncInterfaces shouldn't have a .NETCoreApp TFM. Its netstandard2.1 assembly only has typeforwards, there's no real implementation for .NETCoreApp as all the types are inbox. We shouldn't touch these TFMs.

ViktorHofer reviewed (changes requested):
The TFM changes look incorrect.

sbomer committed:
Fix custom ILLink logic to support running older ILLink

When using downlevel ILLink to trim individual libraries for warning validation,
we shouldn't use custom ILLink options that were introduced in later versions.

sbomer committed:
Avoid multitargeting shim libraries

sbomer committed:
Fix whitespace

ericstj commented on src/libraries/Microsoft.Bcl.AsyncInterfaces/src/Microsoft.Bcl.AsyncInterfaces.csproj:
```diff
@@ -2,6 +2,8 @@
   <PropertyGroup>
     <TargetFrameworks>netstandard2.0;$(NetFrameworkMinimum);netstandard2.1</TargetFrameworks>
+    <!-- Use live ILLink for all TFMs instead of multitargeting. -->
+    <_RequiresLiveILLink>true</_RequiresLiveILLink>
```
> UseLiveILLink

That's better, make sure to include in the comment why it's needed.

> This PR is changing the default so that the custom ILLink logic is only used for NetCoreAppCurrent

Keep in mind that the NetCoreAppCurrent implementation of these Microsoft.BCL libraries is just type forwards.  Not all that interesting...

sbomer replied:
Kept the shims as IsAotCompatible so that we continue running trim analysis on any new code.

sbomer committed:
_RequiresLiveILLink -> UseLiveILLink

sbomer committed:
Set IsAotCompatible false for XmlSerializer.Generator

sbomer committed:
Update comments

sbomer committed:
Set IsAotCompatible false for Diagnostics.EventLog.Messages

sbomer commented on src/libraries/System.Diagnostics.EventLog/src/Messages/System.Diagnostics.EventLog.Messages.csproj:
```diff
@@ -1,6 +1,8 @@
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
+    <!-- Use live ILLink for all TFMs instead of multitargeting. -->
+    <_RequiresLiveILLink>true</_RequiresLiveILLink>
```
Removed this, looks like it's not something that will ever be referenced by the user so it doesn't need to be trimmable or aot compatible.

sbomer committed:
Fix CombineLinkerXmlFiles

sbomer committed:
Merge remote-tracking branch 'origin/main' into illinkAnalyzerOob

# Conflicts:
#	eng/liveILLink.targets
